File: src/server/sourceRoutes.ts
// Routes for an item's sources: its mirrors, checking its PDF URL and mirrors, and rebuilding
// a PDF the store has lost.
import { Semaphore } from "async-mutex";
import type { Context, Hono } from "hono";
import { MirrorRequestSchema, type RebuildOutcome } from "../contract/library";
import { recoverable } from "./indexExport";
import { apiError, invalid, type Library, type LibraryState, now, parseBody } from "./library";
import type { IndexedItem } from "./libraryIndex";
import { addMirror, recordSourceChecks, removeMirror } from "./organization";
import { checkSource, type DownloadSettings, rebuildItem } from "./sources";

export function sourceRoutes(
  app: Hono,
  state: LibraryState,
  root: string,
  settings: DownloadSettings,
  library: Library,
) {
  const downloads = new Semaphore(settings.concurrent_downloads);
  const unknownItem = (c: Context, key: string) =>
    apiError(c, 404, "unknown_item", `no stored PDF has key ${key}`);

  // Checks the PDF URL and every mirror, each download taking a slot of the shared limit.
  const verify = async (indexed: IndexedItem) => {
    const { key, provenance } = indexed.stored;
    const check = (url: string) =>
      downloads.runExclusive(() => checkSource(url, provenance.original_sha256, settings));
    const filing = (await state.organizations.read()).items[key];
    const mirrors = filing === undefined ? [] : filing.mirrors.map((mirror) => mirror.url);
    const [sourceCheck, mirrorChecks] = await Promise.all([
      check(provenance.pdf_url),
      Promise.all(mirrors.map(async (url) => [url, await check(url)] as const)),
    ]);
    const byUrl = new Map(mirrorChecks);
    await state.organizations.update((org) =>
      recordSourceChecks(org, key, provenance.captured_at, sourceCheck, byUrl),
    );
  };

  app.post("/api/items/:key/verify", async (c) => {
    const key = c.req.param("key");
    const indexed = await state.indexed(key);
    if (indexed === undefined) {
      return unknownItem(c, key);
    }
    await verify(indexed);
    return c.json(await state.payloadOf(await state.organizations.read()));
  });

  app.post("/api/verify", async (c) => {
    await Promise.all((await state.index.items()).map(verify));
    return c.json(await state.payloadOf(await state.organizations.read()));
  });

  app.post("/api/items/:key/mirrors", async (c) => {
    const key = c.req.param("key");
    const body = await parseBody(c, MirrorRequestSchema);
    if (!body.success) {
      return invalid(c, body.error);
    }
    if (!(await state.isStored(key))) {
      return unknownItem(c, key);
    }
    return state.change(c, (org) => addMirror(org, key, body.data.url, now()));
  });

  app.delete("/api/items/:key/mirrors", async (c) => {
    const key = c.req.param("key");
    const url = c.req.query("url");
    const filing = (await state.organizations.read()).items[key];
    if (url === undefined || !filing?.mirrors.some((mirror) => mirror.url === url)) {
      return apiError(c, 404, "unknown_mirror", `item ${key} has no mirror ${url}`);
    }
    return state.change(c, (org) => removeMirror(org, key, url, now()));
  });

  const rebuild = async (key: string): Promise<RebuildOutcome | null> => {
    const found = (await state.missing()).find((entry) => entry.item.key === key);
    if (found === undefined) {
      return null;
    }
    const outcome = await rebuildItem(root, recoverable(found.item), settings);
    if (outcome.status === "restored") {
      library.stored();
    }
    return outcome;
  };

  app.post("/api/items/:key/rebuild", async (c) => {
    const key = c.req.param("key");
    const outcome = await rebuild(key);
    if (outcome === null) {
      return apiError(c, 404, "unknown_item", `the index export lists no lost PDF with key ${key}`);
    }
    return c.json(outcome);
  });

  app.post("/api/rebuild", async (c) => {
    const keys = (await state.missing()).map(({ item }) => item.key);
    const outcomes = await Promise.all(
      keys.map((key) => downloads.runExclusive(() => rebuild(key))),
    );
    return c.json(outcomes);
  });
}

File: server/src/source_routes.rs
//! Routes for an item's sources: its mirrors, checking its PDF URL and mirrors, and rebuilding a
//! PDF the store has lost.
use std::collections::{BTreeMap, HashMap};

use axum::body::Bytes;
use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use axum::routing::post;
use axum::{Json, Router};
use futures::future::{join_all, try_join_all};
use url::Url;

use crate::contract::{
    ApiErrorErrorKind, ExportedItem, LibraryPayload, MirrorRequest, RebuildOutcome, Timestamp,
};
use crate::error::{AppError, AppResult};
use crate::export::recoverable;
use crate::index::IndexedItem;
use crate::organization::{add_mirror, filing, record_source_checks, remove_mirror};
use crate::sources::{check_source, rebuild_item};
use crate::state::{parse_body, Shared};

type Payload = AppResult<Json<LibraryPayload>>;

async fn checked(state: &Shared, url: &str, original_sha256: &str) -> crate::contract::SourceCheck {
    let _slot = state
        .downloads
        .acquire()
        .await
        .expect("the semaphore stays open");
    check_source(url, original_sha256, &state.config.app.rebuild).await
}

// Checks the PDF URL and every mirror, each download taking a slot of the shared limit.
async fn verify_item(state: &Shared, indexed: &IndexedItem) -> AppResult<()> {
    let key = indexed.stored.key.to_string();
    let provenance = &indexed.stored.provenance;
    let mirrors: Vec<String> = match filing(&state.organizations.read().await?, &key) {
        Some(filed) => filed
            .mirrors
            .iter()
            .map(|mirror| mirror.url.clone())
            .collect(),
        None => Vec::new(),
    };
    let (source_check, mirror_checks) = futures::join!(
        checked(state, &provenance.pdf_url, &provenance.original_sha256),
        join_all(mirrors.iter().map(|url| async {
            (
                url.clone(),
                checked(state, url, &provenance.original_sha256).await,
            )
        })),
    );
    let by_url: BTreeMap<String, _> = mirror_checks.into_iter().collect();
    let captured_at = provenance.captured_at.clone();
    state
        .organizations
        .update(|org| record_source_checks(org, &key, &captured_at, source_check, &by_url))
        .await?;
    Ok(())
}

async fn verify(State(state): State<Shared>, Path(key): Path<String>) -> Payload {
    let indexed = state.require(&key).await?;
    verify_item(&state, &indexed).await?;
    Ok(Json(state.payload().await?))
}

async fn verify_all(State(state): State<Shared>) -> Payload {
    let indexed = state.store.items().await?;
    try_join_all(indexed.iter().map(|item| verify_item(&state, item))).await?;
    Ok(Json(state.payload().await?))
}

async fn add(State(state): State<Shared>, Path(key): Path<String>, body: Bytes) -> Payload {
    let request: MirrorRequest = parse_body(&body)?;
    match Url::parse(&request.url) {
        Ok(url) if matches!(url.scheme(), "http" | "https") => {}
        Ok(_) => {
            return Err(AppError::invalid(format!(
                "{} is not an http or https URL",
                request.url
            )));
        }
        Err(error) => return Err(AppError::invalid(format!("{}: {error}", request.url))),
    }
    state.require(&key).await?;
    state
        .change(|org| add_mirror(org, &key, &request.url, &Timestamp::now()))
        .await
}

fn unknown_mirror(key: &str, url: &str) -> AppError {
    AppError::api(
        StatusCode::NOT_FOUND,
        ApiErrorErrorKind::UnknownMirror,
        format!("item {key} has no mirror {url}"),
    )
}

async fn remove(
    State(state): State<Shared>,
    Path(key): Path<String>,
    Query(query): Query<HashMap<String, String>>,
) -> Payload {
    let org = state.organizations.read().await?;
    let mirrored = |url: &String| {
        filing(&org, &key)
            .is_some_and(|filed| filed.mirrors.iter().any(|mirror| &mirror.url == url))
    };
    let url = match query.get("url") {
        Some(url) if mirrored(url) => url.clone(),
        Some(url) => return Err(unknown_mirror(&key, url)),
        None => return Err(unknown_mirror(&key, "(none named)")),
    };
    state
        .change(|org| remove_mirror(org, &key, &url, &Timestamp::now()))
        .await
}

/// The items the index export holds whose PDF the store has lost, read once.
async fn lost(state: &Shared) -> AppResult<Vec<ExportedItem>> {
    let indexed = state.store.items().await?;
    Ok(state
        .missing(&indexed)
        .await?
        .into_iter()
        .map(|(item, _)| item)
        .collect())
}

// Rebuilds one lost item; a restored PDF rewrites the index export.
async fn rebuild_lost(state: &Shared, item: &ExportedItem) -> RebuildOutcome {
    let outcome = rebuild_item(&state.store, &recoverable(item), &state.config.app.rebuild).await;
    if let RebuildOutcome::Restored { .. } = outcome {
        state.stored();
    }
    outcome
}

async fn rebuild(
    State(state): State<Shared>,
    Path(key): Path<String>,
) -> AppResult<Json<RebuildOutcome>> {
    let lost = lost(&state).await?;
    let Some(item) = lost.iter().find(|item| *item.key == key) else {
        return Err(AppError::api(
            StatusCode::NOT_FOUND,
            ApiErrorErrorKind::UnknownItem,
            format!("the index export lists no lost PDF with key {key}"),
        ));
    };
    Ok(Json(rebuild_lost(&state, item).await))
}

/// Every lost item's outcome, each download taking a slot of the shared limit; one item's
/// failure is that item's outcome and leaves the others'.
async fn rebuild_all(State(state): State<Shared>) -> AppResult<Json<Vec<RebuildOutcome>>> {
    let lost = lost(&state).await?;
    let outcomes = join_all(lost.iter().map(|item| async {
        let _slot = state
            .downloads
            .acquire()
            .await
            .expect("the semaphore stays open");
        rebuild_lost(&state, item).await
    }))
    .await;
    Ok(Json(outcomes))
}

pub fn routes() -> Router<Shared> {
    Router::new()
        .route("/api/items/{key}/verify", post(verify))
        .route("/api/verify", post(verify_all))
        .route("/api/items/{key}/mirrors", post(add).delete(remove))
        .route("/api/items/{key}/rebuild", post(rebuild))
        .route("/api/rebuild", post(rebuild_all))
}
